# Data Wrangling

Medicare Durable Medical Equipment, Devices & Supplies dataset (via Data.CMS.gov)

Source - [https://data.cms.gov/resources/medicare-durable-medical-equipment-devices-supplies-by-geography-and-service-data-dictionary]

In this section, I will be exploring the raw dataset for any null values, duplicates, and formatting issues. The transformed data will be saved and used in the next section for further analysis.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [2]:
df = pd.read_csv("mup_dme_ry23_p05_v10_dy21_geor.csv")

C:\Users\veron\AppData\Local\Temp\ipykernel_19024\748110536.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("mup_dme_ry23_p05_v10_dy21_geor.csv")


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40267 entries, 0 to 40266
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Rfrg_Prvdr_Geo_Lvl        40267 non-null  object 
 1   Rfrg_Prvdr_Geo_Cd         38567 non-null  object 
 2   Rfrg_Prvdr_Geo_Desc       40267 non-null  object 
 3   BETOS_Lvl                 40267 non-null  object 
 4   BETOS_Cd                  40267 non-null  object 
 5   BETOS_Desc                40267 non-null  object 
 6   HCPCS_Cd                  40267 non-null  object 
 7   HCPCS_Desc                40267 non-null  object 
 8   Suplr_Rentl_Ind           40267 non-null  object 
 9   Tot_Rfrg_Prvdrs           40267 non-null  int64  
 10  Tot_Suplrs                40267 non-null  int64  
 11  Tot_Suplr_Benes           35618 non-null  float64
 12  Tot_Suplr_Clms            40267 non-null  int64  
 13  Tot_Suplr_Srvcs           40267 non-null  int64  
 14  Avg_Su

We can see a 2 columns with some missing values and half of the columns contain strings that will need to be converted into dummy values in a later section. Below we will preview the first 5 rows of the dataset to get an idea of what the data looks like, already we can see some NaN values and strings.

In [4]:
df.head()

Rfrg_Prvdr_Geo_Lvl Rfrg_Prvdr_Geo_Cd Rfrg_Prvdr_Geo_Desc  \
0           National               NaN            National   
1           National               NaN            National   
2           National               NaN            National   
3           National               NaN            National   
4           National               NaN            National   

                        BETOS_Lvl BETOS_Cd              BETOS_Desc HCPCS_Cd  \
0  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4034   
1  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4035   
2  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4036   
3  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4081   
4  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4082   

                                          HCPCS_Desc Suplr_Rentl_Ind  \
0  Enteral feeding supply kit; syringe fed, per d...               N   
1  Enteral feeding supply kit; pump fed, per day,...               N   
2  Enteral feeding supply kit; gravity fed, per d...               N   
3                     Nasogastric tubing with stylet               N   
4                  Nasogastric tubing without stylet               N   

   Tot_Rfrg_Prvdrs  Tot_Suplrs  Tot_Suplr_Benes  Tot_Suplr_Clms  \
0            28183        1745          33451.0          171371   
1            27342        1758          39110.0          271836   
2             6166         933           7171.0           41181   
3               14          12             13.0              20   
4               16          16             16.0              21   

   Tot_Suplr_Srvcs  Avg_Suplr_Sbmtd_Chrg  Avg_Suplr_Mdcr_Alowd_Amt  \
0          4627387             58.266520                  3.718512   
1          7050743             75.495042                  6.433860   
2          1116382             75.172639                  5.036933   
3               33            113.736061                 17.292727   
4               43             99.110930                 12.532093   

   Avg_Suplr_Mdcr_Pymt_Amt  Avg_Suplr_Mdcr_Stdzd_Amt  
0                 2.918066                  3.047039  
1                 5.067509                  5.552222  
2                 3.959289                  4.100195  
3                13.833939                 13.579697  
4                 9.502326                  9.463488

In [5]:
df.dtypes

Rfrg_Prvdr_Geo_Lvl           object
Rfrg_Prvdr_Geo_Cd            object
Rfrg_Prvdr_Geo_Desc          object
BETOS_Lvl                    object
BETOS_Cd                     object
BETOS_Desc                   object
HCPCS_Cd                     object
HCPCS_Desc                   object
Suplr_Rentl_Ind              object
Tot_Rfrg_Prvdrs               int64
Tot_Suplrs                    int64
Tot_Suplr_Benes             float64
Tot_Suplr_Clms                int64
Tot_Suplr_Srvcs               int64
Avg_Suplr_Sbmtd_Chrg        float64
Avg_Suplr_Mdcr_Alowd_Amt    float64
Avg_Suplr_Mdcr_Pymt_Amt     float64
Avg_Suplr_Mdcr_Stdzd_Amt    float64
dtype: object

In [6]:
df.nunique()

Rfrg_Prvdr_Geo_Lvl              2
Rfrg_Prvdr_Geo_Cd              61
Rfrg_Prvdr_Geo_Desc            61
BETOS_Lvl                       3
BETOS_Cd                       12
BETOS_Desc                     12
HCPCS_Cd                     1538
HCPCS_Desc                   1538
Suplr_Rentl_Ind                 2
Tot_Rfrg_Prvdrs              2864
Tot_Suplrs                   1309
Tot_Suplr_Benes              4322
Tot_Suplr_Clms               6196
Tot_Suplr_Srvcs             11815
Avg_Suplr_Sbmtd_Chrg        39686
Avg_Suplr_Mdcr_Alowd_Amt    37303
Avg_Suplr_Mdcr_Pymt_Amt     39863
Avg_Suplr_Mdcr_Stdzd_Amt    39557
dtype: int64

In [7]:
df.describe()

Tot_Rfrg_Prvdrs    Tot_Suplrs  Tot_Suplr_Benes  Tot_Suplr_Clms  \
count     40267.000000  40267.000000     3.561800e+04    4.026700e+04   
mean        447.229692     89.038443     1.730871e+03    4.184466e+03   
std        3937.726657    638.494963     2.451964e+04    6.643068e+04   
min           1.000000      1.000000     1.100000e+01    1.100000e+01   
25%          15.000000      7.000000     2.900000e+01    3.400000e+01   
50%          45.000000     17.000000     8.800000e+01    1.210000e+02   
75%         170.000000     47.000000     3.550000e+02    5.860000e+02   
max      233762.000000  43511.000000     1.891407e+06    6.414962e+06   

       Tot_Suplr_Srvcs  Avg_Suplr_Sbmtd_Chrg  Avg_Suplr_Mdcr_Alowd_Amt  \
count     4.026700e+04          40267.000000              40267.000000   
mean      1.099592e+05            611.400398                361.336364   
std       4.009421e+06           2026.956852               1294.628007   
min       1.100000e+01              0.002468                  0.002468   
25%       5.700000e+01             18.723226                  8.881884   
50%       3.190000e+02             94.654804                 47.792544   
75%       3.378000e+03            426.535897                213.777152   
max       7.279081e+08          49184.117692              38073.511538   

       Avg_Suplr_Mdcr_Pymt_Amt  Avg_Suplr_Mdcr_Stdzd_Amt  
count             40267.000000              40267.000000  
mean                284.765768                279.181098  
std                1024.518576               1003.249283  
min                   0.002222                  0.002222  
25%                   6.811482                  6.713250  
50%                  36.605000                 36.863939  
75%                 168.102759                166.523904  
max               30458.810769              28455.966923

In [8]:
df.describe().T

count           mean           std        min  \
Tot_Rfrg_Prvdrs           40267.0     447.229692  3.937727e+03   1.000000   
Tot_Suplrs                40267.0      89.038443  6.384950e+02   1.000000   
Tot_Suplr_Benes           35618.0    1730.870796  2.451964e+04  11.000000   
Tot_Suplr_Clms            40267.0    4184.465567  6.643068e+04  11.000000   
Tot_Suplr_Srvcs           40267.0  109959.164676  4.009421e+06  11.000000   
Avg_Suplr_Sbmtd_Chrg      40267.0     611.400398  2.026957e+03   0.002468   
Avg_Suplr_Mdcr_Alowd_Amt  40267.0     361.336364  1.294628e+03   0.002468   
Avg_Suplr_Mdcr_Pymt_Amt   40267.0     284.765768  1.024519e+03   0.002222   
Avg_Suplr_Mdcr_Stdzd_Amt  40267.0     279.181098  1.003249e+03   0.002222   

                                25%         50%          75%           max  
Tot_Rfrg_Prvdrs           15.000000   45.000000   170.000000  2.337620e+05  
Tot_Suplrs                 7.000000   17.000000    47.000000  4.351100e+04  
Tot_Suplr_Benes           29.000000   88.000000   355.000000  1.891407e+06  
Tot_Suplr_Clms            34.000000  121.000000   586.000000  6.414962e+06  
Tot_Suplr_Srvcs           57.000000  319.000000  3378.000000  7.279081e+08  
Avg_Suplr_Sbmtd_Chrg      18.723226   94.654804   426.535897  4.918412e+04  
Avg_Suplr_Mdcr_Alowd_Amt   8.881884   47.792544   213.777152  3.807351e+04  
Avg_Suplr_Mdcr_Pymt_Amt    6.811482   36.605000   168.102759  3.045881e+04  
Avg_Suplr_Mdcr_Stdzd_Amt   6.713250   36.863939   166.523904  2.845597e+04

Based on the statistics of the dataset, we can conclude the following:

* Average Referring Providers - 447
* Average Suppliers - 89
* Average Charges - $644.40

In [9]:
df.shape

(40267, 18)

We will now check which columns have missing values and what kinds of values are in these columns. The 2 columns with missing values are Rfrg_Prvdr_Geo_Cd and Tot_Suplr_Benes.

In [10]:
df.isnull().sum()

Rfrg_Prvdr_Geo_Lvl             0
Rfrg_Prvdr_Geo_Cd           1700
Rfrg_Prvdr_Geo_Desc            0
BETOS_Lvl                      0
BETOS_Cd                       0
BETOS_Desc                     0
HCPCS_Cd                       0
HCPCS_Desc                     0
Suplr_Rentl_Ind                0
Tot_Rfrg_Prvdrs                0
Tot_Suplrs                     0
Tot_Suplr_Benes             4649
Tot_Suplr_Clms                 0
Tot_Suplr_Srvcs                0
Avg_Suplr_Sbmtd_Chrg           0
Avg_Suplr_Mdcr_Alowd_Amt       0
Avg_Suplr_Mdcr_Pymt_Amt        0
Avg_Suplr_Mdcr_Stdzd_Amt       0
dtype: int64

Rfrg_Prvdr_Geo_Cd contains categorical values and we will use the mode to fill in the missing values.

In [11]:
df['Rfrg_Prvdr_Geo_Cd'].value_counts()

6.0     1135
36.0    1056
48      1053
12.0    1043
42.0    1005
        ... 
66        45
9D        39
9C        37
69        18
9A         3
Name: Rfrg_Prvdr_Geo_Cd, Length: 61, dtype: int64

In [12]:
df['Rfrg_Prvdr_Geo_Cd'].fillna((df['Rfrg_Prvdr_Geo_Cd'].mode()[0]), inplace=True)

Double checking the Rfrg_Prvdr_Geo_Cd column to make sure the # of rows now match the dataset.

In [13]:
df['Rfrg_Prvdr_Geo_Cd']

0        6.0
1        6.0
2        6.0
3        6.0
4        6.0
        ... 
40262     9E
40263     9E
40264     9E
40265     9E
40266     9E
Name: Rfrg_Prvdr_Geo_Cd, Length: 40267, dtype: object

Tot_Suplr_Benes contains numerical values and we will use the mean to fill in missing values.

In [14]:
df['Tot_Suplr_Benes'].value_counts()

11.0      822
12.0      736
13.0      688
15.0      607
14.0      603
         ... 
961.0       1
5913.0      1
3687.0      1
1252.0      1
1026.0      1
Name: Tot_Suplr_Benes, Length: 4322, dtype: int64

In [15]:
df['Tot_Suplr_Benes'].fillna((df['Tot_Suplr_Benes'].mean()), inplace=True)

Double checking the Tot_Suplr_Benes column to make sure the # of rows now match the dataset.

In [16]:
df['Tot_Suplr_Benes']

0        33451.000000
1        39110.000000
2         7171.000000
3           13.000000
4           16.000000
             ...     
40262       39.000000
40263       12.000000
40264       27.000000
40265       31.000000
40266     1730.870796
Name: Tot_Suplr_Benes, Length: 40267, dtype: float64

Using the isnull() method to double check if our dataset still contains missing values. There are no missing values in our dataset now,

In [17]:
df.isnull().sum()

Rfrg_Prvdr_Geo_Lvl          0
Rfrg_Prvdr_Geo_Cd           0
Rfrg_Prvdr_Geo_Desc         0
BETOS_Lvl                   0
BETOS_Cd                    0
BETOS_Desc                  0
HCPCS_Cd                    0
HCPCS_Desc                  0
Suplr_Rentl_Ind             0
Tot_Rfrg_Prvdrs             0
Tot_Suplrs                  0
Tot_Suplr_Benes             0
Tot_Suplr_Clms              0
Tot_Suplr_Srvcs             0
Avg_Suplr_Sbmtd_Chrg        0
Avg_Suplr_Mdcr_Alowd_Amt    0
Avg_Suplr_Mdcr_Pymt_Amt     0
Avg_Suplr_Mdcr_Stdzd_Amt    0
dtype: int64

Using the duplicated() method to check if our dataset contains any duplicated rows. From the results, there were no duplicated rows in the dataset.

In [18]:
duplicateRowsDF=df[df.duplicated()]
duplicateRowsDF

Empty DataFrame
Columns: [Rfrg_Prvdr_Geo_Lvl, Rfrg_Prvdr_Geo_Cd, Rfrg_Prvdr_Geo_Desc, BETOS_Lvl, BETOS_Cd, BETOS_Desc, HCPCS_Cd, HCPCS_Desc, Suplr_Rentl_Ind, Tot_Rfrg_Prvdrs, Tot_Suplrs, Tot_Suplr_Benes, Tot_Suplr_Clms, Tot_Suplr_Srvcs, Avg_Suplr_Sbmtd_Chrg, Avg_Suplr_Mdcr_Alowd_Amt, Avg_Suplr_Mdcr_Pymt_Amt, Avg_Suplr_Mdcr_Stdzd_Amt]
Index: []

In [19]:
df.shape

(40267, 18)

In [20]:
df.head()

Rfrg_Prvdr_Geo_Lvl Rfrg_Prvdr_Geo_Cd Rfrg_Prvdr_Geo_Desc  \
0           National               6.0            National   
1           National               6.0            National   
2           National               6.0            National   
3           National               6.0            National   
4           National               6.0            National   

                        BETOS_Lvl BETOS_Cd              BETOS_Desc HCPCS_Cd  \
0  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4034   
1  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4035   
2  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4036   
3  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4081   
4  Drugs and Nutritional Products      O1C  Enteral and parenteral    B4082   

                                          HCPCS_Desc Suplr_Rentl_Ind  \
0  Enteral feeding supply kit; syringe fed, per d...               N   
1  Enteral feeding supply kit; pump fed, per day,...               N   
2  Enteral feeding supply kit; gravity fed, per d...               N   
3                     Nasogastric tubing with stylet               N   
4                  Nasogastric tubing without stylet               N   

   Tot_Rfrg_Prvdrs  Tot_Suplrs  Tot_Suplr_Benes  Tot_Suplr_Clms  \
0            28183        1745          33451.0          171371   
1            27342        1758          39110.0          271836   
2             6166         933           7171.0           41181   
3               14          12             13.0              20   
4               16          16             16.0              21   

   Tot_Suplr_Srvcs  Avg_Suplr_Sbmtd_Chrg  Avg_Suplr_Mdcr_Alowd_Amt  \
0          4627387             58.266520                  3.718512   
1          7050743             75.495042                  6.433860   
2          1116382             75.172639                  5.036933   
3               33            113.736061                 17.292727   
4               43             99.110930                 12.532093   

   Avg_Suplr_Mdcr_Pymt_Amt  Avg_Suplr_Mdcr_Stdzd_Amt  
0                 2.918066                  3.047039  
1                 5.067509                  5.552222  
2                 3.959289                  4.100195  
3                13.833939                 13.579697  
4                 9.502326                  9.463488

We check the first 5 rows of our dataset again, we can see there are no longer any NaN values. We save this transformed data to a new file 'Wrangled.csv'.

In [21]:
df.to_csv('Wrangled.csv')